### DISCUSSION TITLE
Add ppb to volatile_organic_compounds

### DISCUSSION THREAD

**Discussion by elupus** [home-assistant/architecture#858] [Entity Models] [answered]
Add ppb as a unit of volatile_organic_compounds.

This seem to have beem missed when checks where added for device classes.

We currently have:
- airthings_ble
- google_assistant
- zha

That are using PPB in core integrations for VOC device class. 

Quote from: https://www.pressac.com/insights/what-are-volatile-organic-compounds-vocs-and-why-should-you-be-monitoring-them/
> Because there are so many types of VOCs, they are often grouped together and monitored as TVOCs (Total Volatile Organic Compounds). can be measured in micrograms per cubic meter (µg/m3) of air, milligrams per cubic meter (mg/m3), parts per million (ppm) or parts per billion (ppb)). Less than 0.3 mg/m3 are considered low TVOC concentration levels. And levels between 0.3 mg/m3 to 0.5 mg/m3 are acceptable.

Indicate units expected are ug/m3, mg/m3, ppm, ppb. This is repeated in other articles.

**Comment by frenck** [comment]
> This seem to have beem missed when checks where added for device classes.

It hasn't been missed. PPB was never supported. All supported units for each device class has always been documented, in both developer and end-user documentation. Integration now carrying this unit with the voc device class, always have been incorrect (and missed during review).

My main concern would be: As far as I understand, we can't provide unit conversion between the existing unit and the proposed unit.

**Reply by elupus** [reply]
No, and neither can the integrations since it's non simple conversion. They all likely lack information from the device to perform any conversion.

So if we can't have this in the device class, we can't mark them an VOC values, which they are.

**Reply by frenck** [reply]
We can still call it VOC, we currently can't declare the device class on it.
Multiple integrations have removed the device class from their sensors in the past weeks for that reason.

Anyways, I'm not against adding it. However, I don't know what the implications are for future/current work on the unit conversions. For example, ug/m3 and mg/m3 would be convertible, and so is ppm and ppb. But I don't think the current model is built to support such cases.

Maybe @emontnemery or @epenet can provide information on that.

**Reply by epenet** [reply]
I agree that if we accept ppb we should also accept ppm and mg/m3.
Regarding conversion the current model won't work so would need to be amended, but maybe conversion for this device_class is not needed? Or could be considered at a later stage.

**Reply by elupus** [reply]
We could standardize on just the two variants for now. ppb and ug/m3, and require integrations to scale to those. Dont think we need conversions here.

**Reply by emontnemery** [reply]
Frenck's concern is valid: µg/m3, mg/m3 and ppm, ppb are really quite different things. µg/m3 and mg/m3 are weight/volume whereas ppm and ppb are unitless ratios (maybe derived from weight/weight?).

Maybe the `volatile_organic_compounds` is not precise enough and should be split in two, one where the reading is a weight/volume and another where the reading is a weight/weight?

**Reply by parautenbach** [reply]
There are formulas to do these conversions, but they depend on temperature. I have these encoded for some REST sensors where I get the values in ppx and convert to density, for example for ozone it is:

```
                {% set temp_kelvin = 273.15 + states('sensor.outside_temperature') | float(0) %}
                {% set const = 12.187 %}
                {% set molecular_weight = 48 %}
                {{ (data.value*const*molecular_weight/temp_kelvin) | round(3) }}
```

But this obviously creates a dependency on temperature, plus it needs to be the temperature where this measurement is taken. This obviously introduces uncertainty and errors into the equation, but for trends and as an indication might still be good enough. That's why I do this for myself, understanding the assumptions and pitfalls, but whether you want to add this by default I don't think so. Probably best for an integration to provide the different measurements itself if it then also has all the right info to do so.

**Reply by epenet** [reply]
If it's a different formula for each compound (ozone, etc.) then it's even more awkward to make it standard.
Maybe two device classes (unitless ratio versus volume ratio) makes sense.

**Reply by parautenbach** [reply]
I know and thought that was clear from the formula – it's based on the molecular weight of each molecule. I was just illustrating that conversation is _possible_ but _error prone_.

**Reply by parautenbach** [reply]
I came across something yesterday I didn't know about, but thought it would be relevant to this discussion. This is useful in particular where unit conversions have to be made under some standard assumptions (if e.g. not all the available information might be available) to make numbers more easily comparable. 

Not great, but the best digest summary I could find is this one (I guess it's not so commonly used): https://www.oxywise.com/what-is-the-difference-between-nm3-and-sm3/. I saw this used with the various particulate matter measurements in an article. 

See also https://en.wikipedia.org/w/index.php?title=Standard_cubic_meters_per_second and https://en.wikipedia.org/wiki/Standard_temperature_and_pressure. 

I've know about standard conditions since school, but I've never seen it used as part of a unit e.g. ug/Nm^3 to indicate it.

**Comment by emontnemery** [comment, marked as answer]
Let's add a new device class, `volatile_organic_compounds_parts` which allows `ppm` and `ppb`

Implemented in https://github.com/home-assistant/core/pull/92973

**Reply by epenet** [reply]
And let's add ppm/ppb to the unit-less ratio converter :)

**Reply by trvrnrth** [reply]
I noticed support was missing from frontend so put together https://github.com/home-assistant/frontend/pull/16576

**Reply by jseidl** [reply]
I know this is a closed thread, but I just realized that ppm/ppb were forgotten to be added as `VALID_UNITS`, only percentage is allowed and it's preventing groups from working for `volatile_organic_compounds_parts`.

See https://github.com/home-assistant/core/blob/dev/homeassistant/util/unit_conversion.py#L660-L673
